src/KUSignalLib/DSP/DSP.py

```diff
@@ -25,4 +26,114 @@ def IIRDirectForm2(b, a, x):
+
+    if dim == 1:  # If the input signal is 1D
+        if interpolate:
+            x_upsampled = Interpolate(x, n, mode="linear")
+        else:
+            for i in range(len(x)):  # Iterate over each element in the input signal
+                x_upsampled += [x[i]] + list(np.zeros(n-1, dtype=type(x[0])))  # Add the current element and n zeros after each element
+        return x_upsampled
+
+    elif dim == 2:  # If the input signal is 2D
+        if axis == 0:  # If axis input is 0, interpolate over rows
+            if interpolate:
+                for i in range(len(x)):
+                    x_upsampled += [list(Interpolate(x[i], n, mode="linear"))]
+            else:
+                for i in range(len(x)):  # Iterate over each row in the input signal
+                    x_i_upsampled = []  # Initialize a list to store the upsampled row
+                    for j in range(len(x[i])):  # Iterate over each element in the current row
+                        x_i_upsampled += [x[i][j]] + list(np.zeros(n-1, dtype=type(x[i][0]))) # Add the current element and n zeros after each element
+                    x_upsampled.append(x_i_upsampled)  # Add the upsampled row to the upsampled signal
+            return x_upsampled
+
+        elif axis == 1:  # If axis input is 1, interpolate over columns
+            if interpolate:
+                pass
+            else:
+                for i in range(len(x)):
+                    for j in range(n-1):
+                        x_upsampled += [x[i]]
+                        for k in range(n-1):
+                            x_upsampled += [list(np.zeros(len(x[i]), dtype=type(x[i][0])))]
+            return x_upsampled
+
+        else:
+            raise ValueError("Invalid axis input.")  # Raise an error indicating invalid axis input
+    else:  # If the input dimension is not 1 or 2
+        raise ValueError("Invalid dimension input.")  # Raise an error indicating invalid dimension input
+
+
+def Downsample(x, n, dim=1, axis=0):
+    """
+    Discrete signal downsample implementation.
+
+    :param x: Input signal.
+    :param n: Downsampled factor.
+    :param dim: Dimension of input signal. Default is 1.
+    :param axis: Axis of x that is downsampled. Default is 0.
+    :return: Downsampled signal.
+    """
+
+    x_downsampled = []  # Initialize an empty list to store the downsampled signal
+    
+    if dim == 1:  # If the dimension is 1D
+        if n > len(x):  # Check if the downsample rate is larger than the signal size
+            raise ValueError("Downsample rate larger than signal size.")
+        
+        # Loop over the signal, downsampling by skipping every n elements
+        for i in range(len(x) // n):
+            x_downsampled.append(x[i*n])
+        
+        return x_downsampled  # Return the downsampled signal
+    
+    elif dim == 2:  # If the dimension is 2D
+        if axis == 0:  # If downsampling over rows
```

Comment:
again, we should down sample over rows and column, user can deal with calling 1d if they just want one or the other